boot/freeldr/freeldr/ntldr/arch/i386/winldr.c

```diff
@@ -381,12 +534,22 @@ WinLdrpMapApic(VOID)
         APICAddress);
 
     /* Map it */
-    HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].PageFrameNumber
-        = APICAddress >> MM_PAGE_SHIFT;
-    HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].Valid = 1;
-    HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].Write = 1;
-    HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].WriteThrough = 1;
-    HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].CacheDisable = 1;
+    if (!PaeModeOn)
+    {
+        HalPageTable[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].PageFrameNumber = APICAddress >> MM_PAGE_SHIFT;
```

Comment:
I'd rather do something like (here and in other places)
```
PHARDWARE_PTE pte = PaeModeOn ? PaeTables->HalPt : HalPageTable;

pte[(APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT].PageFrameNumber = APICAddress >> MM_PAGE_SHIFT;
...
```
Occupies way less space without sacrificing readability

Reply:
We can't use the PHARDWARE_PTE generic structure here because we don't use "#if !defined(_X86PAE_)" 
`#if !defined(_X86PAE_)`
`#define HARDWARE_PTE        HARDWARE_PTE_X86`
`#define PHARDWARE_PTE       PHARDWARE_PTE_X86`
`#else`
`#define HARDWARE_PTE        HARDWARE_PTE_X86_PAE`
`#define PHARDWARE_PTE       PHARDWARE_PTE_X86_PAE`
`#endif`

Therefore, we need to replace all the old (P)HARDWARE_PTE with (P)HARDWARE_PTE_X86!

But it is possible to shorten the code using
`#define APIC_BASE_PTE_IDX   ((APIC_BASE - 0xFFC00000) >> MM_PAGE_SHIFT)`
`#define USER_SHARED_PTE_IDX ((KI_USER_SHARED_DATA - 0xFFC00000) >> MM_PAGE_SHIFT)`
`#define PCR0_PTE_IDX        ((KIP0PCRADDRESS - 0xFFC00000) >> MM_PAGE_SHIFT)`

Reply:
Done